propulsion-hotel/Reactor/GroupCheckoutProcess.fs

```diff
@@ -0,0 +1,53 @@
+module Reactor.GroupCheckoutProcess
+
+open Infrastructure
+open Domain
+
+[<RequireQualifiedAccess>]
+type Outcome = Merged of ok : int * failed : int | Noop
+
+type Service(guestStays : GuestStay.Service, groupCheckouts : GroupCheckout.Service, ?checkoutParallelism) =
+    
+    let checkoutDop = defaultArg checkoutParallelism 5
+    
+    // Attempts a single merge for the specified stay
+    let attemptMerge groupCheckoutId stayId = async {
+        match! guestStays.GroupCheckout(stayId, groupCheckoutId) with
+        | GuestStay.Decide.GroupCheckoutResult.Ok r -> return Choice1Of2 (stayId, r) 
+        | GuestStay.Decide.GroupCheckoutResult.AlreadyCheckedOut -> return Choice2Of2 stayId }
+    
+    // Runs all outstanding merges (with limited parallelism)
+    // Yields residual charges for ones that succeeded,
+    // along with the ids of any that have already been checked out and hence cannot proceed as part of the group checkout
+    let executeMergeStayAttempts groupCheckoutId stayIds = async {
+        let! outcomes = stayIds |> Seq.map (attemptMerge groupCheckoutId) |> Async.parallelLimit checkoutDop
+        return outcomes |> Choice.partition id }
+                
+    // Attempts to merge the specified stays into the specified Group Checkout
+    // Maps the results of each individual merge attempt into 0, 1 or 2 events reflecting the progress achieved against the requested merges 
+    let decideMerge groupCheckoutId stayIds : Async<Outcome * GroupCheckout.Events.Event list> = async {
+        let! residuals, fails = executeMergeStayAttempts groupCheckoutId stayIds
+        let events = [ 
```

Comment:
What happens when the power goes out between these two lines? each of the stays would be marked transfered, but there would be no record of that in the checkout stream. Since transfered is a final state that might pose a problem. Did I miss something?

Reply:
I did miss something, the decider will return the amount if it's for the same checkout

Reply:
So then the question becomes. What if the clerk receives an error, and decides to create a new group checkout in response, can we end up with two groups of checkouts? (I hate thinking about parallelism)

Reply:
Yes, the group checkout owns the work, it won't be forgotten/checked off the list until that has been serviced.
All processing everywhere is and must be idempotent - as you sussed, the retry will kick out the same values next time around.
Assuming no power pulled next time, the event gets written to the group checkout, and the checkpoint moves on.

Regarding the second question re concurrency... The software should probably have an easy way / pre check to highlight group checkouts in progress for that clerk. So when they try to do one they are suggested the current one first.

Failing that, the Stay knows it has been pledged to a given checkout. So perhaps when they pick a stay and its one that's been pledged to a group, but that group one has not been closed, it can navigate you there to instead complete that. Or you can provide a merge group checkout into another group checkout function!

That's where having a reasonable UI flow defined would help more than me making stuff up!

In general though, you absolutely design for this concurrency. In a real world situation the software definitely needs to do something sensible when the inevitable happens and two parties try to do it via two different clerks 5m away from eachother. (Technically a rollback would not be hard)

The equinox-shipping sample covers the returning back of of items if you don't get to reserve all the ones you needed.

Definitely something that's easier to reason about this way than having lots of commands spread around a command queue ;)